R/functions-eager.R

```diff
@@ -1,25 +1,193 @@
-
-  elems <- list2(...)
-
-  if (length(elems) == 0L) {
-    abort("No arguments have been provided to concat.")
-  }
-
-  first <- elems[[1]]
-
-  if (length(elems) == 1L && (is_polars_df(first) || is_polars_series(first) || is_polars_lf(first))) {
-    return(first)
-  }
-
-  if (is_polars_df(first)) {
-    elems |>
-      lapply(\(x) x$`_df`) |>
-      concat_df() |>
-      wrap()
-  } else {
-    abort("todo")
-  }
+# TODO: implement `how = "align"`
+#' Combine multiple DataFrames, LazyFrames, or Series into a single object
+#'
+#' @inheritParams rlang::check_dots_empty0
+#' @param items DataFrames, LazyFrames, or Series to concatenate.
+#' @param how Strategy to concatenate items. Must be one of:
+#' * `"vertical"`: applies multiple vstack operations;
+#' * `"vertical_relaxed"`: same as `"vertical"`, but additionally coerces
+#'   columns to their common supertype if they are mismatched (eg: Int32 to
+#'   Int64);
+#' * `"diagonal"`: finds a union between the column schemas and fills missing
+#'   column values with `null`;
+#' * `"diagonal_relaxed"`: same as `"diagonal"`, but additionally coerces
+#'   columns to their common supertype if they are mismatched (eg: Int32 to
+#'   Int64);
+#' * `"horizontal"`: stacks Series from DataFrames horizontally and fills with
+#'   `null` if the lengths don’t match;
+#' * `"align"`: Combines frames horizontally, auto-determining the common key
+#'   columns and aligning rows using the same logic as `align_frames`; this
+#'   behaviour is patterned after a full outer join, but does not handle
+#'   column-name collision. (If you need more control, you should use a
+#'   suitable join method instead).
+#'
+#' Series only support the `"vertical"` strategy.
+#' @param rechunk Make sure that the result data is in contiguous memory.
+#' @param parallel Only relevant for LazyFrames. This determines if the
+#' concatenated lazy computations may be executed in parallel.
+#'
+#' @return The same class (`polars_data_frame`, `polars_lazy_frame` or
+#' `polars_series`) as the input.
+#' @examples
+#' # default is 'vertical' strategy
+#' df1 <- pl$DataFrame(a = 1L, b = 3L)
+#' df2 <- pl$DataFrame(a = 2L, b = 4L)
+#' pl$concat(c(df1, df2))
+#'
+#' # 'a' is coerced to float64
+#' df1 <- pl$DataFrame(a = 1L, b = 3L)
+#' df2 <- pl$DataFrame(a = 2, b = 4L)
+#' pl$concat(c(df1, df2), how = "vertical_relaxed")
+#'
+#' df_h1 <- pl$DataFrame(l1 = 1:2, l2 = 3:4)
+#' df_h2 <- pl$DataFrame(r1 = 5:6, r2 = 7:8, r3 = 9:10)
+#' pl$concat(c(df_h1, df_h2), how = "horizontal")
+#'
+#' # use 'diagonal' strategy to fill empty column values with nulls
+#' df1 <- pl$DataFrame(a = 1L, b = 3L)
+#' df2 <- pl$DataFrame(a = 2L, c = 4L)
+#' pl$concat(c(df1, df2), how = "diagonal")
+#'
+#' df_a1 <- pl$DataFrame(id = 1:2, x = 3:4)
+#' df_a2 <- pl$DataFrame(id = 2:3, y = 5:6)
+#' df_a3 <- pl$DataFrame(id = c(1L, 3L), z = 7:8)
+#' pl$concat(c(df_a1, df_a2, df_a3), how = "align")
+pl__concat <- function(
+    items,
+    ...,
```

Comment:
In current polars we pass items through dots but if we want to match py-polars we should only pass them via `items`

Reply:
In my opinion, if something like list or dict is the first argument in Python, it should use dynamic dots in R.

- Consistency with functions such as `c()` should also be considered here.
- R lists are not identical to Python lists. There are some tricky points, such as the presence or absence of names, allowing multiple same names, etc. By using dynamic dots, these points can be easily checked and made consistent with many packages.

Reply:
I don't mind that, I updated to use dynamic dots